### PR TITLE
[polynomial] Move primitive root attribute to ntt/intt ops.

### PR DESCRIPTION
**Pull request by j2kun**
Better design to put semantics on the ops, and in this case the ntt/intt op can lower in multiple ways depending on the polynomial ring modulus (it can need an nth root of unity for cyclic polymul -> ntt, or a 2nth root for negacyclic polymul -> ntt)

### CODE REVIEW

File: mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp [hunk 1]
```diff
@@ -157,33 +159,30 @@ static LogicalResult verifyNTTOp(Operation *op, RingAttr ring,
...
-                                 ring.getCoefficientModulus().getValue())) {
-    return op->emitOpError()
-           << "ring type " << ring << " has a primitiveRoot attribute '"
-           << ring.getPrimitiveRoot()
-           << "' that is not a primitive root of the coefficient ring";
+  if (root.has_value()) {
+    APInt rootValue = root.value().getValue().getValue();
+    APInt rootDegree = root.value().getDegree().getValue();
+    APInt cmod = ring.getCoefficientModulus().getValue();
+    if (!isPrimitiveNthRootOfUnity(rootValue, rootDegree, cmod)) {
```
Comment (makslevental) [inline mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp:166]: nit: remove `{`, `}` for single statement if/fors

Comment (makslevental) [reply mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp:166]: maybe it's questionable convention in the case of nested conditionals? up to you.

**Review by makslevental — APPROVED** [review]
LGTM modulo my general distate for algebra (see what I did there...)

File: mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp (hunk 1 above)
Comment (ftynse) [reply mlir/lib/Dialect/Polynomial/IR/PolynomialOps.cpp:166]: It is indeed questionable, I usually lean towards having braces for multi-line bodies even if there is a single statement.

**Review by ftynse — APPROVED** [review]

File: mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td [hunk 2]
```diff
@@ -166,24 +166,45 @@ def Polynomial_RingAttr : Polynomial_Attr<"Ring", "ring"> {
...
         coefficientModulusAttr,
-        polynomialModulusAttr,
-        primitiveRootAttr);
+        polynomialModulusAttr);
     }]>,
   ];
 }
 
+def Polynomial_PrimitiveRootAttr: Polynomial_Attr<"PrimitiveRoot", "primitive_root"> {
+  let summary = "an attribute containing an integer and its degree as a root of unity";
```
[the remark this answers is not in the document]
Comment (j2kun) [reply mlir/include/mlir/Dialect/Polynomial/IR/PolynomialAttributes.td:187]: Looks like yes: 
image

**Pull request merged**